File: api/src/main/java/org/openmrs/logic/Duration.java
/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs.logic;

import org.openmrs.logic.op.ComparisonOperator;
import org.openmrs.logic.op.Operand;

/**
 * Represents a duration of time &mdash; e.g., one year, two weeks, or 18 months. Used within
 * criteria. Use the <code>Duration duration = Duration.days(5.0);</code> to get a duration object
 * 
 * @see org.openmrs.logic.LogicCriteria
 */
public class Duration implements Operand {
	
	public static enum Units {
		SECONDS,
		MINUTES,
		HOURS,
		DAYS,
		WEEKS,
		MONTHS,
		YEARS
	}
	
	private Double duration;
	
	private Units units;
	
	/**
	 * Private contructor used by the static methods on this class. Use the
	 * <code>Duration duration = Duration.days(5.0);</code> to get a duration object
	 * 
	 * @param duration
	 * @param units
	 */
	private Duration(Double duration, Units units) {
		this.duration = duration;
		this.units = units;
	}
	
	public Double getDuration() {
		return duration;
	}
	
	public Units getUnits() {
		return units;
	}
	
	/**
	 * Returns the equivalent duration in days
	 * 
	 * @return the equivalent duration in days
	 */
	public Double getDurationInDays() {
		switch (units) {
			case SECONDS:
				return duration / 86400;
			case MINUTES:
				return duration / 1440;
			case HOURS:
				return duration / 24;
			case DAYS:
				return duration;
			case WEEKS:
				return duration * 7;
			case MONTHS:
				return duration * 30;
			case YEARS:
				return duration * 365;
			default:
				return 0d;
		}
	}
	
	/**
	 * Returns the equivalent duration in milliseconds
	 * 
	 * @return the equivalent duration in milliseconds
	 */
	public long getDurationInMillis() {
		long d = duration.longValue();
		switch (units) {
			case SECONDS:
				return d * 1000;
			case MINUTES:
				return d * 60000;
			case HOURS:
				return d * 3600000;
			case DAYS:
				return d * 86400000;
			case WEEKS:
				return d * 10080000;
			case MONTHS:
				return d * 2628000000L;
			case YEARS:
				return d * 31536000000L;
			default:
				return 0;
		}
	}
	
	/**
	 * Returns a duration for the given number of seconds
	 * 
	 * @param duration number of seconds for duration
	 * @return <code>Duration</code> object for given number of seconds
	 */
	public static Duration seconds(Double duration) {
		return new Duration(duration, Units.SECONDS);
	}
	
	/**
	 * Returns a duration for the given number of seconds
	 * 
	 * @param duration number of seconds for duration
	 * @return <code>Duration</code> object for given number of seconds
	 */
	public static Duration seconds(int duration) {
		return seconds(new Double(duration));
	}
	
	/**
	 * Returns a duration for the given number of minutes
	 * 
	 * @param duration number of minutes for duration
	 * @return <code>Duration</code> object for given number of minutes
	 */
	public static Duration minutes(Double duration) {
		return new Duration(duration, Units.MINUTES);
	}
	
	/**
	 * Returns a duration for the given number of minutes
	 * 
	 * @param duration number of minutes for duration
	 * @return <code>Duration</code> object for given number of minutes
	 */
	public static Duration minutes(int duration) {
		return minutes(new Double(duration));
	}
	
	/**
	 * Returns a duration for the given number of hours
	 * 
	 * @param duration number of hours for duration
	 * @return <code>Duration</code> object for given number of hours
	 */
	public static Duration hours(Double duration) {
		return new Duration(duration, Units.HOURS);
	}
	
	/**
	 * Returns a duration for the given number of hours
	 * 
	 * @param duration number of hours for duration
	 * @return <code>Duration</code> object for given number of hours
	 */
	public static Duration hours(int duration) {
		return hours(new Double(duration));
	}
	
	/**
	 * Returns a duration for the given number of days
	 * 
	 * @param duration number of days for duration
	 * @return <code>Duration</code> object with specified number of days
	 */
	public static Duration days(Double duration) {
		return new Duration(duration, Units.DAYS);
	}
	
	/**
	 * Returns a duration for the given number of days
	 * 
	 * @param duration number of days for duration
	 * @return <code>Duration</code> object with specified number of days
	 */
	public static Duration days(int duration) {
		return days(new Double(duration));
	}
	
	/**
	 * Returns a duration for the given number of weeks
	 * 
	 * @param duration number of weeks for duration
	 * @return <code>Duration</code> object with specified number of weeks
	 */
	public static Duration weeks(Double duration) {
		return new Duration(duration, Units.WEEKS);
	}
	
	/**
	 * Returns a duration for the given number of weeks
	 * 
	 * @param duration number of weeks for duration
	 * @return <code>Duration</code> object with specified number of weeks
	 */
	public static Duration weeks(int duration) {
		return weeks(new Double(duration));
	}
	
	/**
	 * Returns a duration for the given number of months
	 * 
	 * @param duration number of months for duration
	 * @return <code>Duration</code> object with specified number of months
	 */
	public static Duration months(Double duration) {
		return new Duration(duration, Units.MONTHS);
	}
	
	/**
	 * Returns a duration for the given number of months
	 * 
	 * @param duration number of months for duration
	 * @return <code>Duration</code> object with specified number of months
	 */
	public static Duration months(int duration) {
		return months(new Double(duration));
	}
	
	/**
	 * Returns a duration for the given number of years
	 * 
	 * @param duration number of years for duration
	 * @return <code>Duration</code> object with specified number of years
	 */
	public static Duration years(Double duration) {
		return new Duration(duration, Units.YEARS);
	}
	
	/**
	 * Returns a duration for the given number of years
	 * 
	 * @param duration number of years for duration
	 * @return <code>Duration</code> object with specified number of years
	 */
	public static Duration years(int duration) {
		return years(new Double(duration));
	}
	
	/**
	 * @see org.openmrs.logic.op.Operand#supports(org.openmrs.logic.op.ComparisonOperator)
	 */
	public boolean supports(ComparisonOperator operator) {
		return (ComparisonOperator.WITHIN.equals(operator));
	}
	
}

File: api/src/main/kotlin/org/openmrs/Duration.kt
/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs

import org.apache.commons.lang3.time.DateUtils.*
import org.openmrs.api.APIException
import java.util.Date

/**
 * Duration represented using SNOMED CT duration codes
 * 
 * @since 1.10
 */
class Duration(
	private val duration: Int?,
	private val code: String?
) {
	
	/**
	 * Add this duration to given startDate
	 * 
	 * @param startDate
	 * @param frequency is used to calculate time to be added to startDate when duration unit is
	 *            'Recurring Interval'
	 * @return date which is startDate plus duration
	 */
	fun addToDate(startDate: Date, frequency: OrderFrequency?): Date {
		return when (code) {
			SNOMED_CT_SECONDS_CODE -> addSeconds(startDate, duration!!)
			SNOMED_CT_MINUTES_CODE -> addMinutes(startDate, duration!!)
			SNOMED_CT_HOURS_CODE -> addHours(startDate, duration!!)
			SNOMED_CT_DAYS_CODE -> addDays(startDate, duration!!)
			SNOMED_CT_WEEKS_CODE -> addWeeks(startDate, duration!!)
			SNOMED_CT_MONTHS_CODE -> addMonths(startDate, duration!!)
			SNOMED_CT_YEARS_CODE -> addYears(startDate, duration!!)
			SNOMED_CT_RECURRING_INTERVAL_CODE -> {
				if (frequency == null) {
					throw APIException("Duration.error.frequency.null")
				}
				addSeconds(startDate, (duration!! * SECONDS_PER_DAY / frequency.frequencyPerDay).toInt())
			}
			else -> throw APIException("Duration.unknown.code", arrayOf(code))
		}
	}
	
	companion object {
		const val SNOMED_CT_SECONDS_CODE = "257997001"
		const val SNOMED_CT_MINUTES_CODE = "258701004"
		const val SNOMED_CT_HOURS_CODE = "258702006"
		const val SNOMED_CT_DAYS_CODE = "258703001"
		const val SNOMED_CT_WEEKS_CODE = "258705008"
		const val SNOMED_CT_MONTHS_CODE = "258706009"
		const val SNOMED_CT_YEARS_CODE = "258707000"
		const val SNOMED_CT_RECURRING_INTERVAL_CODE = "252109000"
		const val SNOMED_CT_CONCEPT_SOURCE_HL7_CODE = "SCT"
		
		private const val SECONDS_PER_MINUTE = 60
		private const val MINUTES_PER_HOUR = 60
		private const val HOURS_PER_DAY = 24
		private const val SECONDS_PER_HOUR = SECONDS_PER_MINUTE * MINUTES_PER_HOUR
		private const val SECONDS_PER_DAY = SECONDS_PER_HOUR * HOURS_PER_DAY
		
		/**
		 * Returns concept reference term code of the mapping to the SNOMED CT concept source
		 * 
		 * @param durationUnits
		 * @return a string which is reference term code
		 * @should return null if the concept has no mapping to the SNOMED CT source
		 * @should return the code for the term of the mapping to the SNOMED CT source
		 */
		@JvmStatic
		fun getCode(durationUnits: Concept): String? {
			for (conceptMapping in durationUnits.conceptMappings) {
				val conceptReferenceTerm = conceptMapping.conceptReferenceTerm
				if (ConceptMapType.SAME_AS_MAP_TYPE_UUID == conceptMapping.conceptMapType?.uuid &&
					SNOMED_CT_CONCEPT_SOURCE_HL7_CODE == conceptReferenceTerm?.conceptSource?.hl7Code) {
					return conceptReferenceTerm.code
				}
			}
			return null
		}
	}
}
